### PR TITLE
[git] Correctly call `git ls-tree` so unicode files are not quoted

### LINKED ISSUE #3219: Dune segfaults if the codebase contains some UTF-8 chars in (Coq) filenames
Cc @ejgallego @Zimmi48 FYI, I tested the [`dune subst`](https://dune.readthedocs.io/en/latest/usage.html#dune-subst) command (which allows one to automatically add a `version: _` line in `*.opam` files (following the discussion initiated in https://github.com/coq/coq/pull/11038 ))

## Expected Behavior

The command `dune subst` returns normally.

## Actual Behavior

stdout+stderr:

```
Error: exception Sys_error("\"test-suite/misc/deps/\\316\\261\\316\\262/\\316\\263\\316\\264.v\": No such file or directory")
Backtrace:
Raised by primitive operation at file "pervasives.ml", line 385, characters 28-54
Called from file "src/stdune/io.ml", line 87, characters 17-37
Called from file "src/stdune/io.ml" (inlined), line 107, characters 29-64
Called from file "src/watermarks.ml", line 155, characters 10-27
Called from file "list.ml", line 85, characters 12-15
Called from file "src/fiber/fiber.ml", line 201, characters 18-23
Called from file "src/fiber/fiber.ml", line 99, characters 6-13

I must not segfault.  Uncertainty is the mind-killer.  Exceptions are
the little-death that brings total obliteration.  I will fully express
my cases.  Execution will pass over me and through me.  And when it
has gone past, I will unwind the stack along its path.  Where the
cases are handled there will be nothing.  Only I will remain.
```

## Reproduction

1. `git clone https://github.com/coq/coq.git`
1. `cd coq`
1. `dune subst`

## Specifications

- Version of `dune` (output of `dune --version`): 1.11.3
- Version of `ocaml` (output of `ocamlc --version`): 4.05.0
- Operating system (distribution and version): GNU/Linux Debian

## Additional information 

- Link to gist with verbose output (run `dune` with the `--verbose` flag): N/A

- Regular Coq file that triggers that error: [test-suite/misc/deps/αβ/γδ.v](https://github.com/coq/coq/blob/master/test-suite/misc/deps/%CE%B1%CE%B2/%CE%B3%CE%B4.v)

### PR DESCRIPTION
**Pull request by ejgallego**
In order to avoid quoting we need to call `git ls-tree` with the `-z`
option.

This fixes problems with `dune subst` in the presence of unicode
files, in particular fixes #3219

I did a quick-and-dirty implementation of `input_zero_separated`, as I assume we would like to have it elsewhere?

Let me know what you think folks and I can finish it.

Also, I didn't add a test case as I'm a bit scared about unicode filenames in the tree indeed.

### CODE REVIEW

File: src/stdune/io.ml [hunk 1]
```diff
@@ -17,6 +17,38 @@ let input_lines =
...
+
+let input_zero_separated =
+  let buf_size = 65536 in
+  let buf = Bytes.create buf_size in
+  let rec process_buf buf pos acc =
+    match input_zero_from_buffer pos buf with
+    | Some s, eos -> process_buf buf eos (s :: acc)
+    | None, rest -> (acc, rest)
+  in
+  let rec loop ic pos acc =
```
Comment (nojb) [inline src/stdune/io.ml:34]: This code looks buggy. If one of the strings has length `>= Bytes.length buf`, this code will keep reading the string until `buf` gets filled up (ie `pos = buf_size`) and then the exit clause will trigger `res = 0` prematurely.

Either we need to fail if we come across a long string or you need to keep some kind of variable-length buffer to store the string that is currently being read.

**Review by nojb — COMMENTED** [review]
Looks good!

- There is a small problem with the string-reading code
- For the test perhaps you can create the file at runtime to avoid having to commit it

File: src/stdune/io.ml (hunk 1 above)
Comment (ejgallego) [reply src/stdune/io.ml:34]: Oh indeed the code doesn't support a stringer longer than `>= Bytes.length buf`; that's OK for filenames but not in general.

As I mentioned I hacked something quick and dirty, but surely there has to be a reference implementation we could use [maybe in `Base`?] instead of rolling our own, right? This kind of stuff can be tricky (specially perf-wise) so I prefer to use a tested implementation.

Otherwise I'd be happy to do a pass on the code a fix the bugs and maybe tweak a few things here.

File: src/dune_engine/vcs.ml [hunk 2]
```diff
@@ -142,5 +151,5 @@ let files =
 
               let to_dyn = Dyn.Encoder.list Path.to_dyn
             end ))
-       ~git:(f [ "ls-tree"; "-r"; "--name-only"; "HEAD" ])
+       ~git:(f_zero [ "ls-tree"; "-z"; "-r"; "--name-only"; "HEAD" ])
        ~hg:(f [ "files" ])
```
Comment (nojb) [inline src/dune_engine/vcs.ml:155]: `hg` should be given a similar treatment with `hg files -0`.

File: src/stdune/io.ml (hunk 1 above)
Comment (nojb) [reply src/stdune/io.ml:34]: I think it is OK to be simple here. You can look at the implementation of `input_line` for inspiration; there they keep a list of fixed-size buffers instead of resizing a single buffer to fit the string being read.
https://github.com/ocaml/ocaml/blob/10364b67d5d7af71b8626c805ffc0ec49d9b1522/stdlib/stdlib.ml#L439-L465

Comment (ejgallego) [reply src/stdune/io.ml:34]: They do rely on `caml_input_scan_line` tho, we don't have anything equivalent so I'm not sure how we can simplify.

File: src/dune_engine/vcs.ml (hunk 2 above)
Comment (ejgallego) [reply src/dune_engine/vcs.ml]: Done.

File: src/stdune/io.ml (hunk 1 above)
Comment (ejgallego) [reply src/stdune/io.ml]: Ok, I did a new, I hope simpler and more correct version, can you have a look while I add the test?

**Comment by ejgallego** [comment]
Test added, verified it fails on master.

File: src/stdune/io.ml [hunk 3]
```diff
@@ -17,6 +17,47 @@ let input_lines =
...
+      | None -> List.rev acc
+    else
+      (* new input, append remainder and scan it *)
+      let actual_input = Bytes.sub_string ibuf ~pos:0 ~len:res in
+      let actual_input =
+        match rem with
+        | None -> actual_input
+        | Some rem -> rem ^ actual_input
+      in
+      let acc, rem = scan_inputs_buf 0 actual_input acc in
```
Comment (nojb) [inline src/stdune/io.ml:56]: Nitpicking, but I would return `rem, acc` instead of `acc, rem` so as to maintain the same order as in the function calls.

File: src/stdune/io.ml [hunk 4]
```diff
@@ -17,6 +17,47 @@ let input_lines =
...
+      match rem with
+      | Some rem -> List.rev (rem :: acc)
+      | None -> List.rev acc
+    else
+      (* new input, append remainder and scan it *)
+      let actual_input = Bytes.sub_string ibuf ~pos:0 ~len:res in
+      let actual_input =
+        match rem with
+        | None -> actual_input
+        | Some rem -> rem ^ actual_input
```
Comment (nojb) [inline src/stdune/io.ml:54]: If we wanted to improve the worst-case scenario of repeatedly concatenating a string here, `rem` could be a `string list` instead (the function `scan_inputs_buf` would have to be adapted in turn).

But I understand that this is not relevant for the case at hand of dealing with filenames.

**Review by nojb — APPROVED** [review]
LGTM.

I left a few comments, but they are very minor; up to you to take them or not. Thanks!

File: src/stdune/io.ml [hunk 5]
```diff
@@ -17,6 +17,47 @@ let input_lines =
   in
   fun ic -> loop ic []
 
+let input_zero_from_buffer from buf =
+  match String.index_from_opt buf from '\x00' with
+  | None -> (None, from)
+  | Some eos -> (Some (String.sub buf ~pos:from ~len:(eos - from)), eos + 1)
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply src/stdune/io.ml:23]: Done.

File: src/stdune/io.ml (hunk 3 above)
Comment (ejgallego) [reply src/stdune/io.ml:56]: Done.

File: src/stdune/io.ml (hunk 4 above)
Comment (ejgallego) [reply src/stdune/io.ml:54]: Let's improve this indeed.

Comment (ejgallego) [reply src/stdune/io.ml:59]: After giving it a go IMHO the extra complexity of the code was not worth it; I've added a comment instead. We can improve it later if we want.

Comment (nojb) [reply src/stdune/io.ml:57]: Makes sense, thanks.

**Comment by nojb** [comment]
Merged, thanks!
